staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go

```diff
@@ -170,7 +192,15 @@ func (o *APIResourceOptions) RunAPIResources() error {
 		errs = append(errs, err)
 	}
 
-	resources := []groupResource{}
+	rl := &metav1.APIResourceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "v1",
+			APIVersion: "APIResourceList",
+		},
+	}
+	for _, r := range lists {
+		rl.APIResources = append(rl.APIResources, r.APIResources...)
+	}
 
 	for _, list := range lists {
```

Comment:
Shouldn't this entire for loop be contained within `APIResourcesPrinter`?

It only applies to when you are not printing JSON or YAML, right?

Then I think that solves the `resources` global variable problem, because resources can remain a local variable inside `APIResourcesPrinter` now, if I am understanding this correctly.

Reply:
This is actually the part that took me the longest while working on this issue. So I hope I understood things correctly.

So `lists` populated by the call
```go
lists, err := o.discoveryClient.ServerPreferredResources()
```
is a slice of type `metav1.APIResourceList`. Initially, I intended to use it directly in the respective `PrintObj` methods. But while `metav1.APIResourceList` itself implements the `runtime.Object` interface, a slice of it doesn't. Hence, it is how you see it right now.

I could possibly be missing something here. So please correct me if I am wrong. :)

Reply:
OK, I think I understand the reason.

You're re-purposing `metav1.APIResourceList` so that you have something that implements `runtime.Object` for the printer interface to work.

However, this doesn't seem like a good idea to me.  It is really a list of lists, so merging them all into a single APIResourceList is losing some necessary details. For example, I notice in the unit test for json, it is missing the group version (`v1` for Foo and Bar, and `somegroup/v1` for Baz and NoVerbs).